bap-vibes/src/isel.ml

```diff
@@ -665,36 +665,45 @@ module Merge = struct
       congruent = List.append vir1.congruent vir2.congruent
     }
   let merge
-    (ins_outs_map : Data.ins_outs Tid.Map.t)
     (good_templates : Ir.t list) : Ir.t =
     let vir = List.reduce ~f:merge_ir good_templates in
     let vir = Option.value ~default:Ir.empty vir in
-    let vir = Ir.map_blks vir ~f:(fun blk ->
-      match Tid.Map.find ins_outs_map blk.id with
-      | None -> blk
-      | Some {ins ; outs} ->
-        let operands vars =
-          let vars = Var.Set.to_list vars in
-          (* TODO: Assuming Ir.Var is not right here. They may Void. *)
-          let vars = List.map ~f:(fun v -> Ir.Var (Ir.simple_var v)) vars in
-          vars
-        in
-        let in_operands = operands ins in
-        let ins = Ir.empty_op () in
-        let ins = {ins with lhs = blk.ins.lhs @ in_operands} in
-        let out_operands = operands outs in
-        let outs = Ir.empty_op () in
-        let outs = {outs with operands = blk.outs.operands @ out_operands} in
-        {blk with ins; outs}
-      )
-    in
     vir
 
+
 end
 
+let populate_ins_outs (ins_outs_map : Data.ins_outs Tid.Map.t) (vir : Ir.t) : Ir.t =
+  let vir = Ir.map_blks vir ~f:(fun blk ->
+    match Tid.Map.find ins_outs_map blk.id with
+    | None -> blk
+    | Some {ins ; outs} ->
+      let operands vars =
+        let vars = Var.Set.to_list vars in
+        (* TODO: This is not a good way to find register class *)
+        let vars = List.map vars ~f:(fun v ->
+          let sv = Ir.simple_var v in
+          match (Var.typ v) with
+          | Imm _     -> Ir.Var sv
+          | Mem (_,_) -> Ir.Void sv
+          | Unk       -> Ir.Void sv)
+        in
+        vars
+      in
+      let in_operands = operands ins in
+      let ins = Ir.empty_op () in
+      let ins = {ins with lhs = blk.ins.lhs @ in_operands} in
```

Comment:
Is there a chance of duplicates being populated here?

Reply:
I suppose there is a chance. The only intended way that blk.ins.lhs has anything in it is from internal variables in Isel Ir templates on the new isel code path. I don't immediately see a better way of doing this

Reply:
Fair enough, perhaps just a comment here that serves as a footnote for this edge case.